lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1595,6 +1649,169 @@ class BuildBBRegs : public FuncOpPartialLoweringPattern {
   }
 };
 
+/// Builds registers for each pipeline stage in the program.
+class BuildPipelineRegs : public FuncOpPartialLoweringPattern {
+  using FuncOpPartialLoweringPattern::FuncOpPartialLoweringPattern;
+
+  LogicalResult
+  PartiallyLowerFuncToComp(mlir::FuncOp funcOp,
+                           PatternRewriter &rewriter) const override {
+    funcOp.walk([&](staticlogic::PipelineRegisterOp op) {
+      // Condition registers are handled in BuildWhileGroups.
+      auto *parent = op->getParentOp();
+      auto stage = dyn_cast<staticlogic::PipelineStageOp>(parent);
+      if (!stage)
+        return;
+
+      // Create a register for each stage.
+      for (size_t i = 0; i < op.getNumOperands(); ++i) {
+        // Iter args are created in BuildWhileGroups, so just mark the iter arg
+        // register as the appropriate pipeline register.
+        Value stageResult = stage.getResult(i);
+        bool isIterArg = false;
+        for (auto &use : stageResult.getUses()) {
+          if (auto term =
+                  dyn_cast<staticlogic::PipelineTerminatorOp>(use.getOwner())) {
+            if (use.getOperandNumber() < term.iter_args().size()) {
+              WhileOpInterface whileOp(stage->getParentOp());
+              auto reg = getComponentState().getWhileIterReg(
+                  whileOp, use.getOperandNumber());
+              getComponentState().addPipelineReg(stage, reg, i);
```

Comment:
I chose to keep the register declaration for the loop's iter args as it already was, even though they are just like all the other pipeline registers. This has two minor benefits: they get already get initialized correctly, just like scf.while, and they are named to indicate they are iter args. Otherwise they are the same as any pipeline register, and the difference ends in this bit of code.

Reply:
Makes sense to me